text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+- Feature Name: (fill me in with a unique ident, my_awesome_feature)
+- Start Date: 2018-04-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to add futures to libcore, in order to support
+the first-class `async`/`await` syntax proposed in a [companion RFC]. To start
+with, we add the smallest fragment of the futures-rs library required, but we
+anticipate follow-up RFCs ultimately bringing most of the library into libcore (to
+provide a complete complement of APIs).
+
+The proposed APIs are based on the futures crate, but with two major changes:
+
+- The use of [pinned types] to enable borrowing within futures.
+- Removing the associated `Error` type (and adjusting combinators accordingly),
+  in favor of just using `Item = Result<T, E>` instead. The RFC includes an
+  extension trait to provide conveniences for `Result`-producing futures as
+  well.
+
+[pinned types]: https://github.com/rust-lang/rfcs/pull/2349
+[companion RFC]: https://github.com/rust-lang/rfcs/pull/2394
+
+# Motivation
+[motivation]: #motivation
+
+There are two reasons to consider bringing futures into the standard library.
+
+The first, and by far most important, is to provide a supporting mechanism for
+`async`/`await` syntax:
+
+```rust
+async fn function(argument: &str) -> usize {
+     // ...
+}
+```
+
+The syntax itself is motivated in the [companion RFC]. As with closures, it
+involves producing an anonymous type, so that the above declaration is
+equivalent to:
+
+```rust
+fn function<'a>(argument: &'a str) -> _Anonymous<'a, usize> {
+     // ...
+}
+```
+
+and, again like a closure, the type is only usable through the trait it
+implements: `Future`. Hence, to include this syntax in the language, we must
+also introduce `Future` into `core` as a lang item.
+
+The second reason to introduce futures is to establish them more formally as
+*the* way to express composable, asynchronous computation in Rust. Over time,
+the futures library has been refined to a core that bakes in almost no
+assumptions and is usable in a wide variety of contexts, including operating
+systems and embedded devices. As such, it has also become a *de facto* standard
+in this space, and there is a strong desire for the core parts of the library to
+reach the same level of stability as `std` itself. Bringing futures into libcore
+is a final step in standardization.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `Future` trait represents an *asynchronous* computation that may eventually
+produce a final value, but don't have to block the current thread to do so.
```

Comment:
The single most common futures-rs beginner mistake is to assume that a task is spawned in the background as soon as a future is created. So right from the introduction, we may want to stress the fact that Rust's futures are executed lazily, and not eagerly as in other languages.

Here is a possible wording tweak that subtly stresses this point: "...an *asynchronous* and lazy computation...".

Also, there is a small don't -> doesn't typo in there.

Reply:
Would it make sense to add `#[must_use]` then?

Reply:
I definitely think so. I cannot think of any valid use case for creating a future and dropping it afterwards without having done anything else with it, and it is a common mistake.